src/filters/ExtendedKalmanFilter.h

```diff
@@ -0,0 +1,284 @@
+{
+public:
+	using state_t = Eigen::Matrix<double, numStates, 1>;
+	using output_t = Eigen::Matrix<double, numOutputs, 1>;
+	using input_t = Eigen::Matrix<double, numInputs, 1>;
+	using statefunc_t = std::function<state_t(
+		const state_t &, const input_t &, const Eigen::Matrix<double, processNoiseDim, 1> &)>;
+	using outputfunc_t = std::function<output_t(
+		const state_t &, const Eigen::Matrix<double, outputNoiseDim, 1> &)>;
+	using processnoise_t = Eigen::Matrix<double, processNoiseDim, 1>;
+	using outputnoise_t = Eigen::Matrix<double, outputNoiseDim, 1>;
+
+	/**
+	 * Create a new discrete-time EKF.
+	 *
+	 * @param stateFunc Discrete-time state transition function. x_t+1 = f(x_t, u)
+	 * @param outputFunc Output function of the system. Also known as h(x)
+	 * @param processNoise The process noise covariance matrix for this system.
+	 * @param outputNoise The output noise covariance matrix for this system.
+	 * @param dt The time in seconds between updates.
+	 */
+	ExtendedKalmanFilter(
+		const statefunc_t &stateFunc, const outputfunc_t &outputFunc,
+		const NoiseCovMat<numStates, processNoiseDim, numInputs> &processNoise,
+		const NoiseCovMat<numStates, outputNoiseDim, numOutputs> &outputNoise, double dt)
+		: stateFunc(stateFunc), outputFunc(outputFunc), Q(processNoise), R(outputNoise), dt(dt)
+	{
+	}
+
+	void predict(const Eigen::Matrix<double, numInputs, 1> &input) override
+	{
+		Eigen::Matrix<double, numStates, numStates> F =
+			getStateFuncJacobianX(stateFunc, this->xHat, input);
+		Eigen::Matrix<double, processNoiseDim, processNoiseDim> processNoise =
+			Q.get(this->xHat, input);
+
+		Eigen::Matrix<double, numStates, processNoiseDim> L =
+			getStateFuncJacobianW(stateFunc, this->xHat, input);
+
+		this->xHat = stateFunc(this->xHat, input, processnoise_t::Zero());
+		this->P = F * this->P * F.transpose() + L * processNoise * L.transpose();
+	}
+
+	void correct(const Eigen::Matrix<double, numOutputs, 1> &measurement) override
+	{
+		Eigen::Matrix<double, numOutputs, numStates> H =
+			getOutputFuncJacobianX(outputFunc, this->xHat); // output matrix
+
+		Eigen::Matrix<double, outputNoiseDim, outputNoiseDim> outputNoise =
+			R.get(this->xHat, measurement);
+		Eigen::Matrix<double, numOutputs, outputNoiseDim> M =
+			getOutputFuncJacobianV(outputFunc, this->xHat);
+		Eigen::Matrix<double, numOutputs, numOutputs> S =
+			H * this->P * H.transpose() +
+			M * outputNoise * M.transpose(); // residual covariance
+		// near-optimal kalman gain
+		Eigen::Matrix<double, numStates, numOutputs> K =
+			S.transpose().colPivHouseholderQr().solve(H * this->P.transpose()).transpose();
+		output_t y = measurement -
+					 outputFunc(this->xHat, outputnoise_t::Zero()); // measurement residual
+
+		this->xHat = this->xHat + K * y;
+		this->P = (Eigen::Matrix<double, numStates, numStates>::Identity() - K * H) * this->P;
+	}
+
+	/**
+	 * Set this to provide an analytic solution to df/dx.
+	 * If this is null, it will be numerically approximated.
```

Comment:
This is a neat way of doing it.

Some people might think it's a little odd that we provide the state function in the constructor, but specify its derivative by setting a public member variable, but I think this is good enough.

Reply:
To make sure I'm understanding correctly, in this PR the state func jacobians use the default behavior of numerical approximation?

Reply:
Yeah, unless the user specifies analytic solutions to the jacobians of the state and output functions, it will default to numerically differentiating the function itself. The user can specify any or all of the 4 jacobian functions.